File: Object Oriented Programming/Project 1 - cpp/Guarana.cpp
#include <iostream>

#include "World.h"
#include "Organism.h"
#include "Plant.h"
#include "Guarana.h"

Guarana::Guarana() : Plant() {
	_strength = 0;
}

Guarana::Guarana(int x, int y, World* World) : Plant(0, x, y, World) {
}

char Guarana::getSymbol() const {
	return _symbol;
}

string Guarana::getName() const {
	return "Guarana";
}

void Guarana::action() {
	int randomNumber = rand() % 100;
	if (randomNumber < _probability) {
		Plant::action();
	}
}

void Guarana::colision(Organism* attacker) {
	if (this->getStrength() > attacker->getStrength()) {
		this->eventKilling(attacker);
		_World->deleteOrganism(attacker);
	}
	else {
		attacker->eventKilling(this);
		attacker->setStrength(attacker->getStrength() + 3);
		string event = this->getName() + ": strength " + attacker->getName() + " + 3 points";
		_World->addEvent(event);
		_World->putOrganismOnPoint(this->getX(), this->getY(), attacker);
		_World->deleteOrganism(this);
		attacker->print();
	}
}

File: Object Oriented Programming/Project 2 - java/plants/Guarana.java
package project2Java.plants;

import project2Java.Plant;
import project2Java.Organism;
import project2Java.Event;
import project2Java.World;
import project2Java.Position;

import java.awt.*;

public class Guarana extends Plant {
    private static final int STRENGTH_INCREASE = 3;
    private static final int STRENGTH = 0;
    private static final int INITIATIVE = 0;

    public Guarana(World world, Position position, int birthTurn) {
        super(Type.GUARANA, world, position, birthTurn, STRENGTH, INITIATIVE);
        setColor(new Color(252, 3, 15));
    }

    @Override
    public String getOrganismType() {
        return "Guarana";
    }

    @Override
    public boolean specialSkillWhenAttacked(Organism attacker, Organism defender) {
        Position tmpPosition = this.getPosition();
        getWorld().deleteOrganism(this);
        attacker.doMove(tmpPosition);
        Event.addEvent(attacker.getOrganismType() + " eats " + this.getOrganismType() + "  and increase its strength by " + Integer.toString(STRENGTH_INCREASE));
        attacker.setStrength(attacker.getStrength() + STRENGTH_INCREASE);
        return true;
    }
}

